test/iterator.js

```diff
@@ -0,0 +1,367 @@
+'use strict';
+
+const { Iterator, iter } = api.common;
+
+const array = [1, 2, 3, 4];
+
+api.metatests.test('new Iterator() on neither Iterator nor Iterable', test => {
+  test.throws(() => {
+    new Iterator(2);
+  }, new TypeError('Base is neither Iterator nor Iterable'));
+  test.end();
+});
+
+api.metatests.test('iter returns an Iterator', test => {
+  const iterator = iter(array);
+  test.assert(iterator instanceof Iterator);
+  test.end();
+});
+
+api.metatests.test('Iterator is Iterable', test => {
+  const iterator = iter(array);
+  let sum = 0;
+  for (const value of iterator) {
+    sum += value;
+  }
+
+  test.strictSame(sum, 10);
+  test.end();
+});
+
+api.metatests.test('Iterator.each', test => {
+  const iterator = iter(array);
+  let sum = 0;
+  iterator.each(value => {
+    sum += value;
+  });
+
+  test.strictSame(sum, 10);
+  test.end();
+});
+
+api.metatests.test('Iterator.forEach', test => {
+  const iterator = iter(array);
+  let sum = 0;
+  iterator.forEach(value => {
+    sum += value;
+  });
+
+  test.strictSame(sum, 10);
+  test.end();
+});
+
+api.metatests.test('Iterator.forEach with thisArg ', test => {
+  const iterator = iter(array);
+  const obj = {
+    sum: 0,
+    fn(value) {
+      this.sum += value;
+    },
+  };
+
+  iterator.forEach(obj.fn, obj);
+
+  test.strictSame(obj.sum, 10);
+  test.end();
+});
+
+api.metatests.test('Iterator.reduce', test => {
+  test.strictSame(iter(array).reduce((acc, current) => acc + current, 0), 10);
+  test.end();
+});
+
+api.metatests.test('Iterator.reduce with no initialValue', test => {
+  test.strictSame(iter(array).reduce((acc, current) => acc + current), 10);
+  test.end();
+});
+
+api.metatests.test(
+  'Iterator.reduce with no initialValue on consumed iterator',
+  test => {
+    const iterator = iter(array);
+    test.throws(() => {
+      iterator.reduce(() => {});
+      iterator.reduce((acc, current) => acc + current);
+    }, new TypeError('Reduce of consumed iterator with no initial value'));
+
+    test.end();
+  }
+);
+
+api.metatests.test('Iterator.map', test => {
+  test.strictSame(
+    iter(array)
+      .map(value => value * 2)
+      .reduce((previous, current) => previous + current),
```

Comment:
Can `metatests` shallow-compare arrays for equality? Using `reduce()` in a unit test for `map()` doesn't sound right, IMHO.